### ISSUE TITLE
Unit class hierarchy inconsistency

### ISSUE THREAD

**Issue by bevans2000** [mars-sim/mars-sim#1197]
**Describe the bug**
A part of #647 the ```Unit``` hierarchy has been shown to not be accurate and the Unit class is populated with methods that are only applicable to some of the subtypes. Hence it creates an inconsistent design/model.
Unit contains methods that are only relevant to entities that can move around; e.g., setBuilding, setSettlement and more complex logic that needed to identify location state.
The reverse is also true in that classes in the ```Structure``` sub-hierarchy have to implement methods that will never be called.
This means that the Unit class has more logic than needed.

**Expected and unexpected behavior**
The solution is to create a new class ```MovableUnit``` that is a sibling on the existing ```Structure``` class that represents Unit that can move. The latter represents Unit that cannot be moved. This give a clean logic structure.
It will allow some properties/methods to be move from the Unit to the new class as well as some default implementation removed in the Structure hierarchy. 
In addition, it can strengthen the use of the ```transfer``` method as this would become abstract on the ```MoveableUnit``` class

**Comment by mokun** [comment]
Yes it's good idea to create Moveable vs. Nonmoveable Unit.

**Comment by bevans2000** [comment]
I think the new Unit hierarchy should be as below. Key changes:

- ```Settlement``` parented off ``Unit```
- ```MovableUnit``` new class for those Units that can move around a ```Settlement```

![Unit hierachy](https://www.plantuml.com/plantuml/png/VP9DKiCm38NtFeMNS8MkMfZk-EcGeQmTLqqrE5QGPQ2tnwNEn0a9kzZwtfFyedKGmn9RfqmTX8qLRPq9GRzw53Mzh8IZbSYWuLl07p8sNA_q0rdZTXHGaBnkGFhlcrkbhYvt4Tq1VPDD2jiCt9CFlrT9Lg7039Z92aGSjE1bF96dfAf9X5fDnm9JMZ_1b-uCNmv9e2-eKkW5-6Wiw3VYT-0_Jt-aJrCxIEvphHYRawHE1PPRNUL3fz9dLSxVW0od5DOON8nrIWxf9gZ_7h9_ePg61nJAKHtRxZf7o77awWuua5-MPs1HluSJMWUp1jF0Tyvu4CFdhHUKypZq_fyKmqpvp3n70QuwOs69xCqCXohoyJzoEkFs8sBNWbzAD2-p6FJ0cIbKcTZkDrL4cQahDRidcn38eX4uZ3PcdTInTJy0)

**Comment by mokun** [comment]
> MovableUnit``` new class for those Units that can move around a ```Settlement```

Sure. I like the concept of separating movable and non-movable units.

**Comment by bevans2000** [comment]
It will help up the Unit class as it has methods not applicable to all subclasses

**Comment by bevans2000** [comment]
So I've thought about this and we need to make the hierarchy reflect the role the different Unit play. Too much generic functionality is in Unit that is not applicable for the subclasses.
![unit hierarchy](https://www.plantuml.com/plantuml/png/bPJ1Rjim38RlUWeVrrS8Yg9T4M03LZIOiTuPcUa8A5GcqUZwzfCNsMQqSDXk2ldn3_bJydqMI38UWuDzbWHUEXyWv-uRapX7B9WEuF7txqyn39Ywsrl6jxixvyvaZo81ZyZIuKz17chn7rawdiwdhDIS0ZRTOunf80R1tBsYV8aUX29_kD6jFByn3fF2qiaQsUW_B0eFEKTF83YiyQBMJjiB27cjQHBDb2QcnrtwSohVjgeRlEGc6y8kPfgycD3vt8AsVNKR7qSA1NXTPb3b5NcCd2MDVYhiIV0gsexz5FSKKFjd4xP_6ZFySmhOw4jPnYsT2n1ZAlDUOFiO0m9tb3-pShcXx6tOQY_udNp0Iw069vLwR2NiXMhGNcYBBMijp7ossfO4d0_5fDMy6pVVZwzn7zS5hVKgRRpZIkmmvSYsGit_JQ9wPcjKm11fRzGk0AD0Uj-oaBnV-qOzHTLoorVcUHHC_GayMkGi1YdtOphy7_blisn_Z7IQADkSvH5KH1LLOlloq8yaL_BEtPV8U0o_0000)
With this approach we explicitly define:

- Units that can move around and hence need a transfer method and is isInXXXX - ```MobileUnit```
- Units belonging to a Settlement - ```OwnedUnit```
- Units that are static at a fixed position - ```StaticUnit```
- Units that can hold MobileUnits - ```UnitHolder```

**Comment by mokun** [comment]
Great to show the difference between MobileUnit, StaticUnit, and PlanetaryEntity.

In near future, we'll need to clarify what Moon does and we may have MoonSurface instead.

**Comment by bevans2000** [comment]
Just been thinking this through overnight. Do the PlanteryObjects needs to be Units at all?

**Comment by mokun** [comment]
> Do the PlanteryObjects needs to be Units at all?

In our current paradigm of *unit container*, I can see the benefit of having MarsSurface as a reasonably useful unit container.

In the past when MarsSurface did not exist, for a person that just stepped outside, we resort to checking if getSettlement() is null. 

However, this pattern cannot apply to vehicle symmetrically. 

If there's no garage space, a vehicle has to stay outside. 

Its unit container (namely, the settlement) would be null but it still belongs to the settlement. This is one big contradiction.

MarsSurface solves the problem partially (but not completely).

**Comment by bevans2000** [comment]
Right, so in the new world I see these problems being solved by introducing a role approach. So only units that belongs to a Settlement would implement the get settlement method. 
This removes the current inconsistency of many Unit methods not being applicable to all the subclasses. It's bad design modelling and I'm on a mission to fix it

**Comment by mokun** [comment]
> introducing a role approach.

that would be awesome !

I don't particularly like the current unit container approach for tracking unit hierarchy and would love to see some changes to that.

**Comment by bevans2000** [comment]
I've uncovered a bad inconsistency in logic during this work.

- A Vehicle parked at a Settlement is classified as InSettlement
- A Person inside that Vehicle is not classified as InSettlement

To make this solution clean, a Person (or MobileUnit in the new world) inherits the status of it's immediate container. It might not sound  big issue but it's an example of the root problem in the current solution; there are ad-hoc functional patches and it's not logical.
Changing it to be correct will have a massive impact on Tasks as many check a Person is in a Settlement. When in fact they really want to check a Person is **inside** settlement.
Will have to preserve the existing approach and then reworked it afterwards otherwise this issue becomes massive.

**Comment by mokun** [comment]
Let's see if I understand how the current paradigm works 

Scenarios:
1. A vehicle is in a garage. A person is insude this vehicle.

2. A vehicle is parked in settlement vicinity. A person is inside this vehicle. 

3. A vehicle is parked somewhere on Mars on a mission. A person is inside this vehicle.

How does that inconsistency affect the person's choice of tasks ?

**Comment by bevans2000** [comment]
> Let's see if I understand how the current paradigm works
> 
> Scenarios:
> 
> 1. A vehicle is in a garage. A person is inside this vehicle.
> 2. A vehicle is parked in settlement vicinity. A person is inside this vehicle.
> 
> How does that inconsistency affect the person's choice of tasks ?
The question is in the above scenarios, is the Person isInSettlement true for the scenarios 1 & 2?
And is isInSettlement means the Person is inside then what's the point of the isInside method.
And this is my point that it's not logical. 

When a Vehicle arrives at  Settlement it is transferred to the Settlement as a parked vehicle and the parent container is Settlement. So this is inconsistent with how a Person is treated.

**Comment by mokun** [comment]
So one possible me solution is to crkanly define a settlement vicinity as a unit. 

This way, when a vehicle arrives and parks outside (not in a garage), both the person and the vehicle are inside ] the vicinity unit container. Case solved.

**Comment by bevans2000** [comment]
Yes could work. I'm not attempting to solve as part of this issue.

**Comment by mokun** [comment]
> I'm not attempting to solve as part of this issue.

Right. It could be solved under this [thread](https://github.com/mars-sim/mars-sim/issues/1366) once we figure out a few more things in conjunction with that in near future.

**Comment by mokun** [comment]
> I wonder if when arriving at a Settlement it should stay on the MarsSurface but just be logged as a Parked Vehicle. 

So let's clearly defined the verbiage of what *parked vehicles* means, vs. *garaged vehicles*, shall we ?

**Comment by mokun** [comment]
> This would then also work for Person when doing EVA at a Settlement or on a Mission, e.g. near a Vehicle.

Great ! I like that it will work for any persons who are doing EVA near a vehicle that's on Mars surface near a settlement or nowhere near a settlement.

**Comment by bevans2000** [comment]
> > I'm not attempting to solve as part of this issue.
> 
> Right. It could be solved under this [thread](https://github.com/mars-sim/mars-sim/issues/1366) once we figure out a few more things in conjunction with that in near future.

That sounds a good place. I think we need to use vicinity whenever the container is Mars Surface

**Comment by mokun** [comment]
@muybien3433 

Let's read up on the changes that @bevans2000 is currently working on under this thread so as to avoid issues. Thanks !

**Comment by bevans2000** [comment]
Tricky, tricky, tricky..... 
I've got a version of the code working using new ```MobileUnit``` and ```FixedUnit```. I wanted to push on with the ```UnitHolder``` but that's proving a step too far as a single PR. To do the latter is going to get a lot bigger change so I'm going to create a PR that covers the 2 new classes. This will have some placeholder code in place that is not ideal but UnitTest pass and simulations appear to work.
It's not an approach I like but as we are early in the 3.9 release I feel we can live with it. 
Then do a separate set of changes to handle UnitHolder.

@mokun What do you think?

**Comment by mokun** [comment]
Sure. 

You may as well commit the entire thing now if you want. 

I suppose @muybien3433'sPR can adjust to your changes afterward.

**Comment by bevans2000** [comment]
Got some more test runs to do but will look to have something on Saturday

**Comment by bevans2000** [comment]
Overlap with #1356

**Comment by mokun** [comment]
So do we want to redefine "settlement vicinity" as a unit container ?

**Comment by bevans2000** [comment]
I'm not sure at the moment. It feels like a hack but going to work on . 
It comes down to parked vehicles on the surface and people doing local EVA
I guess one question is why do we need it?
There is a lot of complex logic deciding the vicinity case. 
The container is where a Mobile unit is physically residing, so a parked vehicle should remain on the surface unless moved into a garage. Loading and unloading has been reworked recently so it is driven off the Parked vehicle concept and not vicinity.
But we need to move to a cleaner solution gradually with small steps not a big bang. I want to focus on cleaning up the unit hierarchy so the impact of rework is smaller. I'm not convinced the planetary entities need to be Units but this is part of #1197

**Comment by bevans2000** [comment]
Before making the change for the next set of changes there are some places we're we are overusing Unit that will cause a problem when we start removing methods out of Unit. I'm going to clean those up first.
Got to take thee changes slowly but the outcome will be much simpler.

**Comment by mokun** [comment]
> we're we are overusing Unit that will cause a problem when we start removing methods out of Unit. I'm going to clean those up first

Sure !

**Comment by mokun** [comment]
> so a parked vehicle should remain on the surface unless moved into a garage. 

To resolve one confusion, it's obvious we'll probably need to have separate methods, one for creating a list of outside parked vehicles/drones, one for creating a list of garaged vehicles/drones.

**Comment by bevans2000** [comment]
I don't think so. The parked vehicles tells us which vehicles are at a settlement. Those vehicles assigned to a building are garaged. To be assigned a building, they are transferred to the Settlement.

**Comment by mokun** [comment]
> To be assigned a building, they are transferred to the Settlement.

Unfortunately, the way it works currently is that a vehicle that parks in the vicinity of a settlement must be transferred to that settlement as well.

Its container unit was changed from Mars Surface to that settlement.

**Comment by bevans2000** [comment]
We only model Units on the mars surface as everything has a Coordinate. So should we as part of this change add future support for rockets and satellites that reside in space?

In addition, I think Surface features should.d be merged with the MarsSurface class as the latter represents the Mars planet.

**Comment by bevans2000** [comment]
The "Vicinity" issue might be able to be solved by using the LocationStateType better. This has the concept of vicinity. I want to make the updating of this as part of the setContainer method so the 2 are always in sync.

**Comment by mokun** [comment]
> using the LocationStateType better. This has the concept of vicinity.

Right.

> I want to make the updating of this as part of the setContainer method so the 2 are always in sync.

SetContainer is actually doing more than just setting container.

@muybien3433 also made some changes to it in Person recently.


> 
>

**Comment by bevans2000** [comment]
> 
> SetContainer is actually doing more than just setting container.
> 
That my friend is part of the problem 😊
But we can clean this up. I am thinking of this pattern for Transfers. Only UnitHolders can be containers not any Unit. The UnitHolder has two methods

- release(MobileUnit)
- accept(MobileUnit)

So these would be implemented by Settlement, MarsSurface, Vehicle, Person and Robot.

Then the transfer method becomes common as

1. currentContainer.release(this)
2. newContainer.accept(this)
3. Decide new locationStateType
4. this.setContainer( newcontainer, locationState)

**Comment by muybien3433** [comment]
> > using the LocationStateType better. This has the concept of vicinity.
> 
> Right.
> 
> > I want to make the updating of this as part of the setContainer method so the 2 are always in sync.
> 
> SetContainer is actually doing more than just setting container.
> 
> @muybien3433 also made some changes to it in Person recently.

Right, really messy stuff. We should split its functionality. I think UnitHolder could be a solution.
